File: js/src/group-index.ts
import { resultErr, resultOk, type Result } from './result.js';
import { canonicalJson } from './v2/crypto/canonical.js';

export const GROUP_INDEX_SCHEMA = 'aun.group.index.v1';
export const GROUP_INDEX_KEY = 'group.index';
export const GROUP_INDEX_SIG_ALG = 'ECDSA-P256-SHA256';

export interface GroupIndexEntry {
  key: string;
  source: string;
  etag: string;
  last_modified: number;
  [key: string]: unknown;
}

export interface GroupIndexMeta {
  type: 'index_meta';
  group_aid: string;
  etag: string;
  last_modified: number;
  schema: string;
  body_hash: string;
  signed_by: string;
  sig_alg: string;
  signature?: string;
  [key: string]: unknown;
}

export interface SignedGroupIndex {
  body: string;
  meta: GroupIndexMeta;
  entries: GroupIndexEntry[];
}

export interface GroupIndexSigner {
  aid: string;
  sign(payload: Uint8Array | string): Promise<Result<{ signature: string }>>;
  verify(payload: Uint8Array | string, signature: string): Promise<Result<{ valid: boolean }>>;
}

export class GroupIndexMetaCache {
  private readonly remote = new Map<string, Record<string, unknown>>();
  private readonly localEtags = new Map<string, string>();
  private readonly stale = new Set<string>();
  private readonly settings = new Map<string, Record<string, unknown>>();
  private readonly entryEtags = new Map<string, Record<string, string>>();

  observeRpcMeta(meta: Record<string, unknown>, options: { localAid: string }): void {
    const groupIndexes = isRecord(meta?.group_indexes) ? meta.group_indexes : null;
    if (!groupIndexes) return;
    const local = String(options.localAid ?? '');
    for (const [groupAid, value] of Object.entries(groupIndexes)) {
      if (!isRecord(value)) continue;
      const key = this.key(local, groupAid);
      const remoteMeta: Record<string, unknown> = {};
      for (const name of ['etag', 'last_modified', 'schema']) {
        if (value[name] !== undefined && value[name] !== null) remoteMeta[name] = value[name];
      }
      this.remote.set(key, remoteMeta);
      const remoteEtag = String(remoteMeta.etag ?? '');
      if (remoteEtag && this.localEtags.get(key) !== remoteEtag) this.stale.add(key);
    }
  }

  markFresh(localAid: string, groupAid: string, options: { etag: string }): void {
    const key = this.key(localAid, groupAid);
    this.localEtags.set(key, String(options.etag ?? ''));
    this.stale.delete(key);
  }

  isStale(localAid: string, groupAid: string): boolean {
    return this.stale.has(this.key(localAid, groupAid));
  }

  remoteMeta(localAid: string, groupAid: string): Record<string, unknown> | null {
    const value = this.remote.get(this.key(localAid, groupAid));
    return value ? { ...value } : null;
  }

  localEtag(localAid: string, groupAid: string): string {
    return this.localEtags.get(this.key(localAid, groupAid)) ?? '';
  }

  cachedSettings(localAid: string, groupAid: string, keys: string[]): Record<string, unknown> | null {
    const value = this.settings.get(this.key(localAid, groupAid)) ?? {};
    if (!keys.every((item) => item in value)) return null;
    return Object.fromEntries(keys.map((item) => [item, value[item]]));
  }

  cachedSettingsByEntries(
    localAid: string,
    groupAid: string,
    keys: string[],
    entries: GroupIndexEntry[],
  ): { cached: Record<string, unknown>; missing: string[] } {
    const key = this.key(localAid, groupAid);
    const value = this.settings.get(key) ?? {};
    const localEntryEtags = this.entryEtags.get(key) ?? {};
    const remoteEntryEtags = Object.fromEntries(entries.map((item) => [String(item.key ?? ''), String(item.etag ?? '')]));
    const cached: Record<string, unknown> = {};
    const missing: string[] = [];
    for (const item of keys) {
      if (item in value && localEntryEtags[item] === remoteEntryEtags[item]) {
        cached[item] = value[item];
      } else {
        missing.push(item);
      }
    }
    return { cached, missing };
  }

  cacheSettings(
    localAid: string,
    groupAid: string,
    settings: Record<string, unknown>,
    options?: { entries?: GroupIndexEntry[]; etag?: string },
  ): void {
    const key = this.key(localAid, groupAid);
    this.settings.set(key, { ...(this.settings.get(key) ?? {}), ...settings });
    if (options?.entries) {
      const nextEntryEtags = { ...(this.entryEtags.get(key) ?? {}) };
      for (const item of options.entries) {
        const entryKey = String(item.key ?? '');
        if (entryKey) nextEntryEtags[entryKey] = String(item.etag ?? '');
      }
      this.entryEtags.set(key, nextEntryEtags);
    }
    if (options?.etag) this.markFresh(localAid, groupAid, { etag: options.etag });
  }

  restore(
    localAid: string,
    groupAid: string,
    record: {
      remote_meta?: Record<string, unknown>;
      local_etag?: string;
      settings?: Record<string, unknown>;
      entry_etags?: Record<string, unknown>;
    },
  ): void {
    const key = this.key(localAid, groupAid);
    if (isRecord(record.remote_meta)) this.remote.set(key, { ...record.remote_meta });
    if (record.local_etag !== undefined && record.local_etag !== null) {
      this.localEtags.set(key, String(record.local_etag ?? ''));
    }
    if (isRecord(record.settings)) this.settings.set(key, { ...(this.settings.get(key) ?? {}), ...record.settings });
    if (isRecord(record.entry_etags)) {
      this.entryEtags.set(key, {
        ...(this.entryEtags.get(key) ?? {}),
        ...Object.fromEntries(Object.entries(record.entry_etags).map(([entryKey, entryEtag]) => [entryKey, String(entryEtag ?? '')])),
      });
    }
    const remoteEtag = String(this.remote.get(key)?.etag ?? '');
    if (remoteEtag && this.localEtags.get(key) !== remoteEtag) this.stale.add(key);
    if (remoteEtag && this.localEtags.get(key) === remoteEtag) this.stale.delete(key);
  }

  private key(localAid: string, groupAid: string): string {
    return `${String(localAid ?? '')}\x00${String(groupAid ?? '')}`;
  }
}

export async function computeGroupIndexBodyHash(entries: GroupIndexEntry[]): Promise<string> {
  return `sha256:${await sha256Hex(entriesBytes(entries))}`;
}

export async function groupIndexEtag(entries: GroupIndexEntry[]): Promise<string> {
  return `"sha256:${await sha256Hex(entriesBytes(entries))}"`;
}

export function groupIndexSigningPayload(meta: Record<string, unknown>, entries: GroupIndexEntry[]): Uint8Array {
  const metaWithoutSignature: Record<string, unknown> = { ...meta };
  delete metaWithoutSignature.signature;
  const lines = [canonicalStringify(metaWithoutSignature)];
  lines.push(...canonicalEntries(entries).map((item) => canonicalStringify(item)));
  return encode(`${lines.join('\n')}\n`);
}

export async function buildSignedGroupIndex(options: {
  groupAid: string;
  entries: GroupIndexEntry[];
  signer: GroupIndexSigner;
  lastModified: number;
  schema?: string;
}): Promise<SignedGroupIndex> {
  const entries = canonicalEntries(options.entries);
  const meta: GroupIndexMeta = {
    type: 'index_meta',
    group_aid: String(options.groupAid),
    etag: await groupIndexEtag(entries),
    last_modified: Math.trunc(Number(options.lastModified)),
    schema: String(options.schema ?? GROUP_INDEX_SCHEMA),
    body_hash: await computeGroupIndexBodyHash(entries),
    signed_by: options.signer.aid,
    sig_alg: GROUP_INDEX_SIG_ALG,
  };
  const signed = await options.signer.sign(groupIndexSigningPayload(meta, entries));
  if (!signed.ok) throw new Error(signed.error.message || 'group index signing failed');
  meta.signature = signed.data.signature;
  const body = [canonicalStringify(meta), ...entries.map((item) => canonicalStringify(item))].join('\n') + '\n';
  return { body, meta, entries };
}

export function parseGroupIndex(body: string | { body?: unknown }): { meta: GroupIndexMeta; entries: GroupIndexEntry[] } {
  const text = typeof body === 'string' ? body : String(body?.body ?? '');
  const lines = text.split(/\r?\n/).filter((line) => line.trim());
  if (lines.length === 0) throw new Error('group index body is empty');
  const meta = JSON.parse(lines[0]) as GroupIndexMeta;
  const entries = lines.slice(1).map((line) => JSON.parse(line) as GroupIndexEntry);
  if (meta.type !== 'index_meta') throw new Error('first group index line must be index_meta');
  return { meta, entries };
}

export async function verifyGroupIndex(
  body: string | { body?: unknown },
  signer: GroupIndexSigner,
): Promise<Result<{ valid: false; reason: string } | { valid: true; meta: GroupIndexMeta; entries: GroupIndexEntry[] }>> {
  try {
    const parsed = parseGroupIndex(body);
    const signature = String(parsed.meta.signature ?? '');
    if (!signature) return resultOk({ valid: false, reason: 'signature missing' });
    if (String(parsed.meta.signed_by ?? '') !== signer.aid) return resultOk({ valid: false, reason: 'signed_by mismatch' });
    if (String(parsed.meta.sig_alg ?? '') !== GROUP_INDEX_SIG_ALG) return resultOk({ valid: false, reason: 'unsupported sig_alg' });
    if (String(parsed.meta.body_hash ?? '') !== await computeGroupIndexBodyHash(parsed.entries)) {
      return resultOk({ valid: false, reason: 'body_hash mismatch' });
    }
    if (String(parsed.meta.etag ?? '') !== await groupIndexEtag(parsed.entries)) {
      return resultOk({ valid: false, reason: 'etag mismatch' });
    }
    const verified = await signer.verify(groupIndexSigningPayload(parsed.meta, parsed.entries), signature);
    if (!verified.ok) return resultErr(verified.error.code, verified.error.message || 'group index verify failed', verified.error.cause);
    if (!verified.data.valid) return resultOk({ valid: false, reason: 'signature verification failed' });
    return resultOk({ valid: true, meta: parsed.meta, entries: canonicalEntries(parsed.entries) });
  } catch (exc) {
    return resultErr('GROUP_INDEX_VERIFY_ERROR', String(exc), exc);
  }
}

export async function prepareGroupSettingsWithIndex(options: {
  groupAid: string;
  settings: Record<string, unknown>;
  signer: GroupIndexSigner;
  lastModified: number;
  baseIndex?: string | { body?: unknown } | null;
}): Promise<Record<string, unknown>> {
  const result: Record<string, unknown> = { ...options.settings };
  const updatedEntries: GroupIndexEntry[] = [];
  for (const [key, value] of Object.entries(options.settings)) {
    if (key !== GROUP_INDEX_KEY) updatedEntries.push(await settingEntry(key, value, options.lastModified));
  }
  const updatedKeys = new Set(updatedEntries.map((item) => item.key));
  const entries: GroupIndexEntry[] = [];
  if (options.baseIndex) {
    const parsed = parseGroupIndex(options.baseIndex);
    entries.push(...parsed.entries.filter((item) => !updatedKeys.has(String(item.key))).map((item) => ({ ...item })));
  }
  entries.push(...updatedEntries);
  result[GROUP_INDEX_KEY] = await buildSignedGroupIndex({
    groupAid: options.groupAid,
    entries,
    signer: options.signer,
    lastModified: options.lastModified,
  });
  return result;
}

async function settingEntry(key: string, value: unknown, lastModified: number): Promise<GroupIndexEntry> {
  const digest = await sha256Hex(encode(canonicalStringify(value)));
  return {
    key,
    source: 'db',
    etag: `"sha256:${digest}"`,
    last_modified: Math.trunc(Number(lastModified)),
  };
}

function canonicalEntries(entries: GroupIndexEntry[]): GroupIndexEntry[] {
  return entries.map((item) => ({ ...item })).sort((a, b) => compareCodePoints(String(a.key ?? ''), String(b.key ?? '')));
}

function entriesBytes(entries: GroupIndexEntry[]): Uint8Array {
  const lines = canonicalEntries(entries).map((item) => canonicalStringify(item));
  return encode(lines.length ? `${lines.join('\n')}\n` : '');
}

function canonicalStringify(value: unknown): string {
  return new TextDecoder().decode(canonicalJson(value));
}

function encode(text: string): Uint8Array {
  return new TextEncoder().encode(text);
}

async function sha256Hex(data: Uint8Array): Promise<string> {
  const digest = await globalThis.crypto.subtle.digest('SHA-256', exactArrayBuffer(data));
  return bytesToHex(new Uint8Array(digest));
}

function exactArrayBuffer(data: Uint8Array): ArrayBuffer {
  if (data.byteOffset === 0 && data.byteLength === data.buffer.byteLength && data.buffer instanceof ArrayBuffer) {
    return data.buffer;
  }
  return data.slice().buffer as ArrayBuffer;
}

function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes).map((b) => b.toString(16).padStart(2, '0')).join('');
}

function compareCodePoints(a: string, b: string): number {
  const ac = Array.from(a);
  const bc = Array.from(b);
  const n = Math.min(ac.length, bc.length);
  for (let i = 0; i < n; i++) {
    const av = ac[i].codePointAt(0) ?? 0;
    const bv = bc[i].codePointAt(0) ?? 0;
    if (av !== bv) return av - bv;
  }
  return ac.length - bc.length;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return !!value && typeof value === 'object' && !Array.isArray(value);
}

File: python/src/aun_core/group_index.py
from __future__ import annotations

import hashlib
import json
import os
from pathlib import Path
from typing import Any
import uuid

from .aid import AID
from .result import Result, result_err, result_ok
from .v2.crypto.canonical import canonical_json as _canonical_json_bytes

GROUP_INDEX_SCHEMA = "aun.group.index.v1"
GROUP_INDEX_KEY = "group.index"
GROUP_INDEX_SIG_ALG = "ECDSA-P256-SHA256"


class GroupIndexMetaCache:
    def __init__(self, aun_path: str | os.PathLike[str] | None = None) -> None:
        self._aun_path = Path(aun_path) if aun_path else None
        self._remote: dict[tuple[str, str], dict[str, Any]] = {}
        self._local_etags: dict[tuple[str, str], str] = {}
        self._stale: set[tuple[str, str]] = set()
        self._settings: dict[tuple[str, str], dict[str, Any]] = {}
        self._entry_etags: dict[tuple[str, str], dict[str, str]] = {}

    def observe_rpc_meta(self, meta: dict[str, Any], *, local_aid: str) -> None:
        group_indexes = meta.get("group_indexes") if isinstance(meta, dict) else None
        if not isinstance(group_indexes, dict):
            return
        local = str(local_aid or "")
        for group_aid, value in group_indexes.items():
            if not isinstance(value, dict):
                continue
            key = (local, str(group_aid))
            remote = {
                name: value.get(name)
                for name in ("etag", "last_modified", "schema")
                if value.get(name) is not None
            }
            self._load_key(key)
            self._remote[key] = remote
            remote_etag = str(remote.get("etag") or "")
            if remote_etag and self._local_etags.get(key) != remote_etag:
                self._stale.add(key)
            self._save_key(key)

    def mark_fresh(self, local_aid: str, group_aid: str, *, etag: str) -> None:
        key = (str(local_aid or ""), str(group_aid or ""))
        self._load_key(key)
        self._local_etags[key] = str(etag or "")
        self._stale.discard(key)
        self._save_key(key)

    def is_stale(self, local_aid: str, group_aid: str) -> bool:
        key = (str(local_aid or ""), str(group_aid or ""))
        self._load_key(key)
        return key in self._stale

    def remote_meta(self, local_aid: str, group_aid: str) -> dict[str, Any] | None:
        key = (str(local_aid or ""), str(group_aid or ""))
        self._load_key(key)
        value = self._remote.get(key)
        return dict(value) if value else None

    def local_etag(self, local_aid: str, group_aid: str) -> str:
        key = (str(local_aid or ""), str(group_aid or ""))
        self._load_key(key)
        return self._local_etags.get(key, "")

    def cached_settings(self, local_aid: str, group_aid: str, keys: list[str]) -> dict[str, Any] | None:
        key = (str(local_aid or ""), str(group_aid or ""))
        self._load_key(key)
        settings = self._settings.get(key) or {}
        if any(item not in settings for item in keys):
            return None
        return {item: settings[item] for item in keys}

    def cached_settings_by_entries(
        self,
        local_aid: str,
        group_aid: str,
        keys: list[str],
        entries: list[dict[str, Any]],
    ) -> tuple[dict[str, Any], list[str]]:
        key = (str(local_aid or ""), str(group_aid or ""))
        self._load_key(key)
        settings = self._settings.get(key) or {}
        local_entry_etags = self._entry_etags.get(key) or {}
        remote_entry_etags = {str(item.get("key") or ""): str(item.get("etag") or "") for item in entries}
        cached: dict[str, Any] = {}
        missing: list[str] = []
        for item in keys:
            if item in settings and local_entry_etags.get(item) == remote_entry_etags.get(item):
                cached[item] = settings[item]
            else:
                missing.append(item)
        return cached, missing

    def cache_settings(
        self,
        local_aid: str,
        group_aid: str,
        settings: dict[str, Any],
        *,
        entries: list[dict[str, Any]] | None = None,
        etag: str = "",
        group_index: str | dict[str, Any] | None = None,
    ) -> None:
        key = (str(local_aid or ""), str(group_aid or ""))
        self._load_key(key)
        current = dict(self._settings.get(key) or {})
        current.update(settings)
        self._settings[key] = current
        if entries is not None:
            entry_etags = dict(self._entry_etags.get(key) or {})
            for item in entries:
                entry_key = str(item.get("key") or "")
                if entry_key:
                    entry_etags[entry_key] = str(item.get("etag") or "")
            self._entry_etags[key] = entry_etags
        if etag:
            self.mark_fresh(local_aid, group_aid, etag=etag)
        else:
            self._save_key(key)
        self._save_group_index_body(key, group_index)

    def _dir_for_key(self, key: tuple[str, str]) -> Path | None:
        if self._aun_path is None:
            return None
        local_aid, group_aid = key
        if not local_aid or not group_aid:
            return None
        for value in key:
            if "/" in value or "\\" in value or "\0" in value:
                raise ValueError("group index cache aid must not contain path separators")
        return self._aun_path / "AIDs" / local_aid / "groups" / group_aid

    def _cache_path_for_key(self, key: tuple[str, str]) -> Path | None:
        cache_dir = self._dir_for_key(key)
        return cache_dir / "group-index-cache.json" if cache_dir is not None else None

    def _index_path_for_key(self, key: tuple[str, str]) -> Path | None:
        cache_dir = self._dir_for_key(key)
        return cache_dir / "index.jsonl" if cache_dir is not None else None

    def _load_key(self, key: tuple[str, str]) -> None:
        if key in self._settings or key in self._local_etags or key in self._remote:
            return
        path = self._cache_path_for_key(key)
        if path is None or not path.exists():
            return
        try:
            data = json.loads(path.read_text(encoding="utf-8"))
        except Exception:
            return
        if isinstance(data.get("remote_meta"), dict):
            self._remote[key] = dict(data["remote_meta"])
        local_etag = str(data.get("local_etag") or "")
        if local_etag:
            self._local_etags[key] = local_etag
        if isinstance(data.get("settings"), dict):
            self._settings[key] = dict(data["settings"])
        if isinstance(data.get("entry_etags"), dict):
            self._entry_etags[key] = {str(k): str(v) for k, v in data["entry_etags"].items()}
        remote_etag = str((self._remote.get(key) or {}).get("etag") or "")
        if remote_etag and self._local_etags.get(key) != remote_etag:
            self._stale.add(key)

    def _save_key(self, key: tuple[str, str]) -> None:
        path = self._cache_path_for_key(key)
        if path is None:
            return
        payload = {
            "local_aid": key[0],
            "group_aid": key[1],
            "remote_meta": self._remote.get(key) or {},
            "local_etag": self._local_etags.get(key, ""),
            "settings": self._settings.get(key) or {},
            "entry_etags": self._entry_etags.get(key) or {},
        }
        path.parent.mkdir(parents=True, exist_ok=True)
        tmp = path.with_name(f".{path.name}.{os.getpid()}.{uuid.uuid4().hex}.tmp")
        tmp.write_text(json.dumps(payload, ensure_ascii=False, indent=2, sort_keys=True) + "\n", encoding="utf-8")
        os.replace(tmp, path)

    def _save_group_index_body(self, key: tuple[str, str], group_index: str | dict[str, Any] | None) -> None:
        if group_index is None:
            return
        body = group_index.get("body") if isinstance(group_index, dict) else group_index
        text = str(body or "")
        if not text:
            return
        path = self._index_path_for_key(key)
        if path is None:
            return
        path.parent.mkdir(parents=True, exist_ok=True)
        tmp = path.with_name(f".{path.name}.{os.getpid()}.{uuid.uuid4().hex}.tmp")
        tmp.write_text(text if text.endswith("\n") else text + "\n", encoding="utf-8")
        os.replace(tmp, path)


def _canonical_json(value: Any) -> str:
    return _canonical_json_bytes(value).decode("utf-8")


def _canonical_entries(entries: list[dict[str, Any]]) -> list[dict[str, Any]]:
    normalized = [dict(item) for item in entries]
    return sorted(normalized, key=lambda item: str(item.get("key") or ""))


def _entries_bytes(entries: list[dict[str, Any]]) -> bytes:
    lines = [_canonical_json(item) for item in _canonical_entries(entries)]
    return ("\n".join(lines) + ("\n" if lines else "")).encode("utf-8")


def compute_group_index_body_hash(entries: list[dict[str, Any]]) -> str:
    return "sha256:" + hashlib.sha256(_entries_bytes(entries)).hexdigest()


def group_index_etag(entries: list[dict[str, Any]]) -> str:
    return '"sha256:' + hashlib.sha256(_entries_bytes(entries)).hexdigest() + '"'


def group_index_signing_payload(meta: dict[str, Any], entries: list[dict[str, Any]]) -> bytes:
    meta_without_signature = dict(meta)
    meta_without_signature.pop("signature", None)
    lines = [_canonical_json(meta_without_signature)]
    lines.extend(_canonical_json(item) for item in _canonical_entries(entries))
    return ("\n".join(lines) + "\n").encode("utf-8")


def build_signed_group_index(
    *,
    group_aid: str,
    entries: list[dict[str, Any]],
    signer: AID,
    last_modified: int,
    schema: str = GROUP_INDEX_SCHEMA,
) -> dict[str, Any]:
    canonical_entries = _canonical_entries(entries)
    meta = {
        "type": "index_meta",
        "group_aid": str(group_aid),
        "etag": group_index_etag(canonical_entries),
        "last_modified": int(last_modified),
        "schema": str(schema),
        "body_hash": compute_group_index_body_hash(canonical_entries),
        "signed_by": signer.aid,
        "sig_alg": GROUP_INDEX_SIG_ALG,
    }
    signed = signer.sign(group_index_signing_payload(meta, canonical_entries))
    if not signed.ok:
        raise ValueError(signed.message or "group index signing failed")
    meta["signature"] = signed.data["signature"]
    body_lines = [_canonical_json(meta)]
    body_lines.extend(_canonical_json(item) for item in canonical_entries)
    body = "\n".join(body_lines) + "\n"
    return {"body": body, "meta": meta, "entries": canonical_entries}


def parse_group_index(body: str | dict[str, Any]) -> dict[str, Any]:
    if isinstance(body, dict):
        body = str(body.get("body") or "")
    lines = [line for line in str(body or "").splitlines() if line.strip()]
    if not lines:
        raise ValueError("group index body is empty")
    meta = json.loads(lines[0])
    entries = [json.loads(line) for line in lines[1:]]
    if meta.get("type") != "index_meta":
        raise ValueError("first group index line must be index_meta")
    return {"meta": meta, "entries": entries}


def verify_group_index(body: str | dict[str, Any], signer: AID) -> Result[dict[str, Any]]:
    try:
        parsed = parse_group_index(body)
        meta = parsed["meta"]
        entries = parsed["entries"]
        signature = str(meta.get("signature") or "")
        if not signature:
            return result_ok({"valid": False, "reason": "signature missing"})
        if str(meta.get("signed_by") or "") != signer.aid:
            return result_ok({"valid": False, "reason": "signed_by mismatch"})
        if str(meta.get("sig_alg") or "") != GROUP_INDEX_SIG_ALG:
            return result_ok({"valid": False, "reason": "unsupported sig_alg"})
        expected_hash = compute_group_index_body_hash(entries)
        if str(meta.get("body_hash") or "") != expected_hash:
            return result_ok({"valid": False, "reason": "body_hash mismatch"})
        expected_etag = group_index_etag(entries)
        if str(meta.get("etag") or "") != expected_etag:
            return result_ok({"valid": False, "reason": "etag mismatch"})
        verified = signer.verify(group_index_signing_payload(meta, entries), signature)
        if not verified.ok:
            return result_err(verified.code, verified.message or "group index verify failed")
        if not bool(verified.data.get("valid")):
            return result_ok({"valid": False, "reason": "signature verification failed"})
        return result_ok({"valid": True, "meta": meta, "entries": _canonical_entries(entries)})
    except Exception as exc:
        return result_err("GROUP_INDEX_VERIFY_ERROR", str(exc), cause=exc)


def _setting_entry(key: str, value: Any, last_modified: int) -> dict[str, Any]:
    value_bytes = _canonical_json(value).encode("utf-8")
    digest = hashlib.sha256(value_bytes).hexdigest()
    return {
        "key": key,
        "source": "db",
        "etag": f'"sha256:{digest}"',
        "last_modified": int(last_modified),
    }


def prepare_group_settings_with_index(
    *,
    group_aid: str,
    settings: dict[str, Any],
    signer: AID,
    last_modified: int,
    base_index: str | dict[str, Any] | None = None,
) -> dict[str, Any]:
    result = dict(settings)
    updated_entries = [
        _setting_entry(key, value, last_modified)
        for key, value in settings.items()
        if key != GROUP_INDEX_KEY
    ]
    updated_keys = {str(item.get("key") or "") for item in updated_entries}
    entries: list[dict[str, Any]] = []
    if base_index:
        parsed = parse_group_index(base_index)
        entries.extend(
            dict(item)
            for item in parsed["entries"]
            if str(item.get("key") or "") not in updated_keys
        )
    entries.extend(updated_entries)
    result[GROUP_INDEX_KEY] = build_signed_group_index(
        group_aid=group_aid,
        entries=entries,
        signer=signer,
        last_modified=last_modified,
    )
    return result
